edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
-                                await AddOrRemoveSubscription(listener, true, subscriptionPattern.Subscrition);
+                                await AddOrRemoveSubscription(listener, true, subscriptionPattern.Subscription);
+
+                                var subscriptions = default(HashSet<DeviceSubscription>);
+                                if (!affectedClients.TryGetValue(identity, out subscriptions))
+                                {
+                                    subscriptions = new HashSet<DeviceSubscription>();
+                                    affectedClients.Add(identity, subscriptions);
+                                }
+
+                                subscriptions.Add(subscriptionPattern.Subscription);
                             }
                             catch (Exception e)
                             {
-                                Events.FailedToChangeSubscriptionState(e, subscriptionPattern.Subscrition.ToString(), identity.Id);
+                                Events.FailedToChangeSubscriptionState(e, subscriptionPattern.Subscription.ToString(), identity.Id);
                             }
                         }
                         else
                         {
-                            Events.CouldNotObtainListener(subscriptionPattern.Subscrition.ToString(), identity.Id);
+                            Events.CouldNotObtainListenerForSubscription(subscriptionPattern.Subscription.ToString(), identity.Id);
                         }
                     }
                 }
             }
 
-            return true;
+            // At this point we have a collection about 'affectedClients'. Those are the clients subscribed to something.
+            // What we don't know are the unsubscribtions. That is because the broker sends only the current subscription list,
+            // so when a client unsubscribes, that is only a shorter list of subscriptions sent.
+            // The information we want to gain is that what the unsubscriptions are, and the answer is that what we have not seen
+            // previously as subscribed, that is unsubscribed. E.g. if a client subscribed to 'methods' only, then we say that
+            // it unsubscribed from twin responses, c2d, m2m, etc. The underlaying layers take care of the optimization doing
+            // nothing when unsubscribed twice from a topic.
+            // As a first step we need all nested connections, because it is possible that the subscribtion list above did not
+            // cover all clients. It can happen when a client had only a single subscription before, but now it unsubscribed.
+            // In that case that client will not be contained by the subscription event sent by the broker, because it has no
+            // subscription at all.
+            var nestConnections = await this.connectionRegistry.GetNestedConnectionsAsync();
```

Comment:
this part till end could also be extracted in a separate method

Reply:
extracted